### ISSUE TITLE
pstate is set on wrong GPU

### ISSUE THREAD

**Issue by mostlygeek** [sasha0552/nvidia-pstated#3]
This bug happens on v1.0.5 and latest commit hash 153c435ed6f2de4b26e18784a0a253ab64ca1b24. 

I've attached a video that shows what's happening more clearly. I have 4 GPUs in my system, 3xP40s and 1x3090. 

In my testing I use `CUDA_VISIBLE_DEVICES` with the uuid of the GPU I'm targeting. Using `nvidia-smi` we can see where inference is happening and that the `pstate` is being changed on the incorrect GPU despite what the logs are outputting. 

https://github.com/user-attachments/assets/910ed7d2-e341-4b1f-8c62-b0e56632f959

**Comment by sasha0552** [comment]
Hi, thanks for the detailed bug report! Could you please provide the list of GPU names (at startup) in the [dev/mismatched-ids](https://github.com/sasha0552/nvidia-pstated/tree/dev/mismatched-ids) branch?

I think this behavior is due to the different ordering in the handles used by NVML (for monitoring) and NVAPI (for management).

**Comment by mostlygeek** [comment]
Hi! Thanks for the quick response. 

Here's the output: 

```
ids =  (0)
iterationsBeforeSwitch = 30
performanceStateHigh = 16
performanceStateLow = 8
sleepInterval = 100
temperatureThreshold = 80
0. Tesla P40 (GPU id = 0)
(NVAPI) 0. NVIDIA GeForce RTX 3090 (GPU id = 0)
1. Tesla P40 (GPU id = 1)
(NVAPI) 1. Tesla P40 (GPU id = 1)
2. Tesla P40 (GPU id = 2)
(NVAPI) 2. Tesla P40 (GPU id = 2)
3. NVIDIA GeForce RTX 3090 (GPU id = 3)
(NVAPI) 3. Tesla P40 (GPU id = 3)
Managing 4 GPUs...
GPU 0 entered performance state 8
GPU 1 entered performance state 8
GPU 2 entered performance state 8
GPU 3 entered performance state 8
```

**Comment by mostlygeek** [comment]
Tried it and get this error now: 

`NvAPI_GPU_GetBoardInfo(nvapiDevices[i], &boardInfo): NVAPI_NOT_SUPPORTED`

**Comment by sasha0552** [comment]
It seems that the 3090 (maybe, since it works on my P40s) doesn't support `NvAPI_GPU_GetBoardInfo`, which is used to get the GPU's serial number to sort handles based on it.

Can you get the serial number (you don't have to post it) using `nvidia-smi --query-gpu=name,serial --format=csv` for the 3090? If it's not available even there, I think I should redo the sorting based on the PCI bus id (since there's no way to get the GPU UUID in NVAPI).

**Comment by sasha0552** [comment]
I replaced the handle identifier with the PCI bus id, please test again.

**Comment by mostlygeek** [comment]
```
$ ./build/nvidia-pstated
ids = N/A (0)
iterationsBeforeSwitch = 30
performanceStateHigh = 16
performanceStateLow = 8
sleepInterval = 100
temperatureThreshold = 80
0. Tesla P40 (GPU id = 0)
(NVAPI) 0. Tesla P40 (GPU id = 0)
1. Tesla P40 (GPU id = 1)
(NVAPI) 1. Tesla P40 (GPU id = 1)
2. Tesla P40 (GPU id = 2)
(NVAPI) 2. Tesla P40 (GPU id = 2)
3. NVIDIA GeForce RTX 3090 (GPU id = 3)
(NVAPI) 3. NVIDIA GeForce RTX 3090 (GPU id = 3)
Managing 4 GPUs...
GPU 0 entered performance state 8
GPU 1 entered performance state 8
GPU 2 entered performance state 8
GPU 3 entered performance state 8
```

I tested using llama-cli and llama-server, with various combinations of single and multi-gpu and it all looks good. When utilization goes up on the specific GPU the pstate is updated. 

Great job!

**Comment by sasha0552** [comment]
Merged into `main` and released as `v1.0.6`.

**Comment by LoopControl** [comment]
Just wanted to say thanks for creating this program and making these fixes. It's a life saver for me with the P40.
